internal/rbd/snap_diff.go

```diff
@@ -0,0 +1,344 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rbd
+
+import (
+	"context"
+	"errors"
+
+	rbderrors "github.com/ceph/ceph-csi/internal/rbd/errors"
+	"github.com/ceph/ceph-csi/internal/util"
+	"github.com/ceph/ceph-csi/internal/util/log"
+
+	librbd "github.com/ceph/go-ceph/rbd"
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	// TODO: Make internalMaxResults configurable via a flag or config option.
+	// internalMaxResults is the maximum number of results to return in a single
+	// response. This is used to limit the number of changed blocks that needs
+	// to be stored in memory before sending a response to the stream.
+	internalMaxResults = 20
+
+	// streamDoneErrCode is a custom error code used to indicate that the
+	// stream was closed by the client.
+	streamDoneErrCode = -1
+	// sendResponseErrCode is a custom error code used to indicate that
+	// sending the response to the stream failed.
+	sendResponseErrCode = -2
+)
+
+// metadataStream is an interface that is used to as a common type for
+// csi.SnapshotMetadata_GetMetadataAllocatedServer and
+// csi.SnapshotMetadata_GetMetadataDeltaServer. This allows us to use a single
+// function to send responses to both types of streams without duplicating code.
+type metadataStream interface{}
+
+// sendStreamResponse sends the response to the stream based on its type.
+// It handles both GetMetadataAllocated and GetMetadataDelta responses.
+// It returns an error if the stream type is invalid or if sending the response fails.
+func sendStreamResponse(
+	ctx context.Context,
+	stream metadataStream,
+	volSize int64,
+	changedBlocks []*csi.BlockMetadata) error {
+	if len(changedBlocks) == 0 {
+		return nil
+	}
+
+	log.DebugLog(ctx, "Sending %d changed blocks to the stream", len(changedBlocks))
+	switch s := stream.(type) {
+	case csi.SnapshotMetadata_GetMetadataAllocatedServer:
+		return s.Send(&csi.GetMetadataAllocatedResponse{
+			BlockMetadataType:   csi.BlockMetadataType_VARIABLE_LENGTH,
+			VolumeCapacityBytes: volSize,
+			BlockMetadata:       changedBlocks,
+		})
+	case csi.SnapshotMetadata_GetMetadataDeltaServer:
+		return s.Send(&csi.GetMetadataDeltaResponse{
+			BlockMetadataType:   csi.BlockMetadataType_VARIABLE_LENGTH,
+			VolumeCapacityBytes: volSize,
+			BlockMetadata:       changedBlocks,
+		})
+	default:
+		return errors.New("invalid stream type")
+	}
+}
+
+func (rbdSnap *rbdSnapshot) streamMetadata(
+	ctx context.Context,
+	stream metadataStream,
+	startingOffset int64,
+	maxResults int32,
+	baseRBDSnap *rbdSnapshot,
+) error {
+	var fromSnapID uint64
+	image, err := rbdSnap.open()
+	if err != nil {
+		return status.Errorf(codes.Internal, "failed to open image %q: %s", rbdSnap, err)
```

Comment:
`rbdSnap` functions should not return gRPC errors. gRPC errors should only be created in `ControllerServer` or other gRPC servers.